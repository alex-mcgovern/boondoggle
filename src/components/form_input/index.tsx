import { useController, useFormContext } from "react-hook-form";

import { Input } from "../input";

import type { WithFormFieldProps } from "../../common-types";
import type { InputProps } from "../input";

/**
 * React Hook Form connected version of Boondoggle's `Input`. Uses `useFormContext`
 * to access Hook Form's methods so can be nested in markup. Must be a descendant of `FormProvider`
 */
export type FormInputProps = InputProps & WithFormFieldProps;

export function FormInput({
    defaultValue,
    invalid,
    name,
    onChange,
    wrapperProps,
    ...rest
}: FormInputProps) {
    const { control } = useFormContext();

    const {
        field: { onBlur, onChange: reactHookFormOnChange, ref, value: controlledValue = "" },
        fieldState: { error },
    } = useController({
        control,
        defaultValue,
        name,
    });

    return (
        <Input
            errorMessage={error?.message}
            invalid={invalid || !!error}
            name={name}
            onBlur={onBlur}
            onChange={(e) => {
                onChange?.(e);

                reactHookFormOnChange(e);
            }}
            ref={ref}
            value={controlledValue}
            wrapperProps={wrapperProps}
            {...rest}
        />
    );
}
