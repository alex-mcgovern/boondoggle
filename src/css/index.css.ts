import { createSprinkles, defineProperties } from "@vanilla-extract/sprinkles";

import { withPrefersMotion } from "../_css-utils";
import {
    MEDIA_QUERY_DESKTOP,
    MEDIA_QUERY_TABLET,
    sprinklesLayer,
    vars,
} from "../index.css";

const border = {
    border_field: `1px solid ${vars.color.border_field}`,
    border_field_active: `1px solid ${vars.color.border_field_active}`,
    border_rule: `1px solid ${vars.color.border_rule}`,
    focus: `1px solid ${vars.color.focus_border} !important`,
    none: "0px solid transparent",
    transparent: "1px solid transparent",
};

const properties = defineProperties({
    "@layer": sprinklesLayer,
    properties: {
        alignItems: ["stretch", "start", "center", "end", "baseline"],
        alignSelf: ["stretch", "start", "center", "end", "baseline"],
        aspectRatio: vars.aspectRatio,
        background: {
            ...vars.color,
            inherit: "inherit",
            transparent: "transparent",
        },
        border: border,
        borderBottom: border,
        borderLeft: border,
        borderRadius: vars.borderRadius,
        borderRight: border,
        borderTop: border,
        bottom: ["0"],
        boxShadow: vars.boxShadow,
        color: {
            ...vars.color,
            currentColor: "currentColor",
            inherit: "inherit",
        },
        cursor: ["pointer", "default", "not-allowed", "grab"],
        flex: ["0 1 auto", "1 1 auto", "1 1 0%", "1 1 100%"],
        flexGrow: ["0", "1"],
        flexShrink: ["0", "1"],
        flexWrap: ["wrap", "nowrap"],
        fontSize: vars.fontSize,
        fontWeight: vars.fontWeight,
        height: { ...vars.height, ...vars.spacing },
        inset: ["0"],
        isolation: ["isolate"],
        justifyContent: [
            "stretch",
            "start",
            "center",
            "end",
            "space-around",
            "space-between",
        ],
        justifySelf: ["stretch", "start", "center", "end", "baseline"],
        left: ["0"],
        lineHeight: vars.lineHeight,
        margin: vars.spacing,
        marginBottom: vars.spacing,
        marginLeft: vars.spacing,
        marginRight: vars.spacing,
        marginTop: vars.spacing,
        maxHeight: { ...vars.height, ...vars.spacing },
        maxWidth: { ...vars.width, ...vars.spacing },
        minHeight: { ...vars.height, ...vars.spacing },
        minWidth: { ...vars.width, ...vars.spacing },
        opacity: ["0", "0.3", "0.5", "0.8", "1"],
        outline: {
            // Group: Default outline should come first to allow overriding
            none: "0px solid transparent",

            // Group: Focus outline should come last to allow overriding
            focus: `1px solid ${vars.color.focus_ring} !important`,
        },
        overflowX: ["hidden", "auto", "visible", "scroll"],
        overflowY: ["hidden", "auto", "visible", "scroll"],
        right: ["0"],
        textAlign: ["center", "left", "right"],
        textDecoration: ["underline", "none", "line-through"],
        textTransform: ["capitalize", "uppercase", "lowercase"],
        textWrap: ["balance"],
        top: ["0"],
        transition: {
            long: withPrefersMotion({
                transitionDuration: vars.transitionDuration.long,
                transitionProperty: "all",
                transitionTimingFunction: vars.ease.quart_in_out,
            }),
            medium: withPrefersMotion({
                transitionDuration: vars.transitionDuration.medium,
                transitionProperty: "all",
                transitionTimingFunction: vars.ease.quart_in_out,
            }),
            short: withPrefersMotion({
                transitionDuration: vars.transitionDuration.short,
                transitionProperty: "all",
                transitionTimingFunction: vars.ease.quart_in_out,
            }),
        },
        whiteSpace: ["nowrap", "normal"],
        zIndex: ["-1", "1"],
    },
    shorthands: {
        fontStyle: ["fontSize", "lineHeight"],
        marginX: ["marginLeft", "marginRight"],
        marginY: ["marginTop", "marginBottom"],
        overflow: ["overflowX", "overflowY"],
        placeItems: ["alignItems", "justifyContent"],
    },
});

const responsiveProperties = defineProperties({
    conditions: {
        desktop: {
            "@media": MEDIA_QUERY_DESKTOP,
        },
        mobile: {},
        tablet: {
            "@media": MEDIA_QUERY_TABLET,
        },
    },
    defaultCondition: "mobile",
    properties: {
        display: vars.display,
        flexDirection: ["row", "column", "row-reverse", "column-reverse"],
        gap: vars.spacing,
        gridTemplateColumns: vars.gridTemplateColumns,
        padding: vars.spacing,
        paddingBottom: vars.spacing,
        paddingLeft: vars.spacing,
        paddingRight: vars.spacing,
        paddingTop: vars.spacing,
        position: ["relative", "absolute", "sticky", "static"],
        width: { ...vars.width, ...vars.spacing },
    },
    shorthands: {
        paddingX: ["paddingLeft", "paddingRight"],
        paddingY: ["paddingTop", "paddingBottom"],
    },
});

export const css = createSprinkles(properties, responsiveProperties);

export type Css = Parameters<typeof css>[0];
