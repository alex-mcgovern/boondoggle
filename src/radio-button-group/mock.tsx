import type * as React from "react";

import type { RadioButton } from "../radio-button";

import { LOREM } from "../../mocks/LOREM.mock";

export const RADIO_BUTTONS_MOCK: Array<
	React.ComponentProps<typeof RadioButton>
> = [
	{
		description: LOREM.text_xxs,
		name: "radio_item_1",
		title: "Radio item 1",
		value: "radio_item_1",
	},
	{
		description: LOREM.text_xxs,
		name: "radio_item_2",
		title: "Radio item 2",
		value: "radio_item_2",
	},
	{
		description: LOREM.text_xxs,
		name: "radio_item_3",
		title: "Radio item 3",
		value: "radio_item_3",
	},
];
