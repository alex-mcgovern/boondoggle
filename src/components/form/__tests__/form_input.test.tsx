/**
 * @jest-environment jsdom
 */
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@vanilla-extract/css/disableRuntimeStyles";

import { LOREM } from "../../../../mocks/LOREM.mock";
import { FormInput } from "../../form_input";
import { FormTestProvider } from "../form_test_provider.comp";

import type { FormInputProps } from "../../form_input";

const PROPS: FormInputProps = {
    errorMessage: LOREM.errorMessage(),
    name: LOREM.name(),
    placeholder: LOREM.placeholder(),
};

const renderComponent = ({ ...props }: FormInputProps) => {
    return render(
        <FormTestProvider>
            <FormInput {...props} />
        </FormTestProvider>
    );
};

describe("<FormInput />", () => {
    describe("Basic smoke tests", () => {
        test("should render without throwing", () => {
            const { getByRole } = renderComponent(PROPS);

            expect(getByRole("textbox")).not.toBeNull();
        });

        test("should match snapshot", () => {
            const { getByRole } = renderComponent(PROPS);

            expect(getByRole("textbox")).toMatchSnapshot();
        });
    });

    describe("Updating the value", () => {
        test("should update value when the user types", async () => {
            const { getByRole } = renderComponent(PROPS);

            expect(getByRole("textbox")).toHaveValue("");

            await userEvent.type(getByRole("textbox"), "New value");

            expect(getByRole("textbox")).toHaveValue("New value");
        });
    });
});
