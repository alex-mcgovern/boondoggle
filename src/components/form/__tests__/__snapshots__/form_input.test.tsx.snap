// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormInput /> Basic smoke tests should match snapshot 1`] = `
<input
  class="input.styles_getInputStyles__3x32wp4 input.styles__3x32wp3 get_sprinkles_width_100%_mobile__eiij7bap a11y_a11yDisabled__r8he5t1 input.styles_getInputStyles_size_md__3x32wp6 input.styles_size_md__3x32wp1 element_size_elementFontSize_md__1f6zrv4 element_size__1f6zrv1 get_sprinkles_fontSize_body_md__eiij7b91 get_sprinkles_lineHeight_body_md__eiij7b9d element_size_elementPadding_md__1f6zrva element_size__1f6zrv7 get_sprinkles_paddingLeft_spacing1.5__eiij7b7z get_sprinkles_paddingRight_spacing1.5__eiij7b8a get_sprinkles_paddingTop_spacing1__eiij7b8k get_sprinkles_paddingBottom_spacing1__eiij7b7n"
  name="Name ipsum"
  placeholder="Placeholder ipsum"
  value=""
/>
`;
